### PR TITLE
add ServiceManager and ServiceBroker - gain control over memory

### PR DESCRIPTION
**Pull request by FernetMenta**
Start to get rid of evil statics and get control over the sequence of crating object.

@garbear @mapfau this should fix the crash on exit.

### CODE REVIEW

File: xbmc/Application.cpp [hunk 1]
```diff
@@ -2846,6 +2839,12 @@ bool CApplication::Cleanup()
     delete m_network;
     m_network = NULL;
 
+    if (m_ServiceManager)
```
Comment (Paxxi) [inline xbmc/Application.cpp]: This should not be needed, deinit should be the destructor unless we specifically need to deinit and reinit during execution

File: xbmc/Application.h [hunk 2]
```diff
@@ -390,7 +390,7 @@ class CApplication : public CXBApplicationEx, public IPlayerCallback, public IMs
    */
   void UnregisterActionListener(IActionListener *listener);
 
-  ADDON::CBinaryAddonCache m_binaryAddonCache;
+  std::shared_ptr<CServiceMgr> m_ServiceManager;
```
Comment (Paxxi) [inline xbmc/Application.h]: seems like this should be a unique_ptr

File: xbmc/ServiceManager.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "ServiceManager.h"
+#include "utils/log.h"
+
+CServiceMgr::~CServiceMgr()
```
Comment (Paxxi) [inline xbmc/ServiceManager.cpp]: this can be removed or replaced with ~CServiceMgr() = default; in the header

File: xbmc/ServiceManager.cpp [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  return true;
+}
+
+void CServiceMgr::Deinit()
+{
+  m_binaryAddonCache.reset();
+  m_addonMgr.reset();
+}
+
+ADDON::CAddonMgr *CServiceMgr::GetAddonMgr()
```
Comment (Paxxi) [inline xbmc/ServiceManager.cpp]: can these ever return nullptr? If not it might make sense to return a reference instead

**Comment by Paxxi** [comment]
I like the idea @FernetMenta , might disagree about some minors but it's probably a good direction to go to try and get rid of all our statics

File: xbmc/Application.cpp (hunk 1 above)
Comment (FernetMenta) [inline xbmc/Application.cpp]: CApplication::Cleanup is called twice, at least on OSX

Comment (Paxxi) [inline xbmc/Application.cpp]: That seems like something that should be fixed but then it's an understandable design.

File: xbmc/ServiceManager.cpp (hunk 4 above)
Comment (FernetMenta) [inline xbmc/ServiceManager.cpp]: I was thinking about placing an assert here. if this is nullptr something went wrong in a way that app won't run

File: xbmc/ServiceManager.cpp (hunk 3 above)
Comment (FernetMenta) [inline xbmc/ServiceManager.cpp]: right

File: xbmc/Application.cpp (hunk 1 above)
Comment (FernetMenta) [inline xbmc/Application.cpp]: I place a comment here for now. Can't fix all bugs at once :)

Comment (Paxxi) [inline xbmc/Application.cpp]: sounds good :)

File: xbmc/ServiceManager.cpp (hunk 4 above)
Comment (tamland) [inline xbmc/ServiceManager.cpp]: then it should be a reference. shouldn't expose a mutable pointer at least

File: xbmc/addons/AddonManager.h [hunk 5]
```diff
@@ -240,9 +246,9 @@ namespace ADDON
     /*! \brief Stop all services addons.
     */
     void StopServices(const bool onlylogin);
-    static AddonPtr Factory(const cp_plugin_info_t* plugin, TYPE type);
-    static AddonPtr Factory(const cp_plugin_info_t* plugin, TYPE type, CAddonBuilder& builder);
-    static void FillCpluffMetadata(const cp_plugin_info_t* plugin, CAddonBuilder& builder);
+    AddonPtr Factory(const cp_plugin_info_t* plugin, TYPE type);
+    AddonPtr Factory(const cp_plugin_info_t* plugin, TYPE type, CAddonBuilder& builder);
+    void FillCpluffMetadata(const cp_plugin_info_t* plugin, CAddonBuilder& builder);
```
Comment (tamland) [inline xbmc/addons/AddonManager.h]: Looks unrelated to this commit. 

it's static for a good reason: the keep it pure. it should not have access to anything internally, only the cpluff helper functions.

File: xbmc/ServiceManager.cpp (hunk 4 above)
Comment (FernetMenta) [inline xbmc/ServiceManager.cpp]: ok

File: xbmc/addons/AddonManager.h (hunk 5 above)
Comment (FernetMenta) [inline xbmc/addons/AddonManager.h]: ok

**Comment by FernetMenta** [comment]
@Paxxi  @tamland updated according to your comments. Mind having another look?

@tamland you may want to revisit AddonMgr at a later time. As you suggested I changed the methods back to static. The call stack looks a bit weird because ServiceManager initializes AddonMgr and its static methods calls back into ServiceManager to get the instance of AddonMgr.

File: xbmc/ServiceManager.cpp [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+void CServiceMgr::Deinit()
+{
+  m_binaryAddonCache.reset();
+  m_addonMgr.reset();
+}
+
+ADDON::CAddonMgr &CServiceMgr::GetAddonMgr()
+{
+  return *m_addonMgr.get();
```
Comment (Paxxi) [inline xbmc/ServiceManager.cpp]: I don't think you need the get here http://en.cppreference.com/w/cpp/memory/unique_ptr/operator* but I either way works so not something you have to change

**Comment by Paxxi** [comment]
Looks good to me but I defer to @tamland for final say as it's addon related

File: xbmc/ServiceManager.h [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+#include "addons/AddonManager.h"
+#include "addons/BinaryAddonCache.h"
+#include <memory>
+
+namespace ADDON {
+class CAddonMgr;
+class CBinaryAddonCache;
+}
+
+class CServiceMgr
```
Comment (Montellese) [inline xbmc/ServiceManager.h]: Could we rename this to `CServiceManager`? I've always found this very confusing about `CAddonMgr` because I have to include `addons/AddonManager.h` and then I expect the class to be named `CAddonManager` like with all other classes where the name of the class matches the name of the header file (prepended with a `C`) but then it doesn't compile because the class name is actually `CAddonMgr`.

File: xbmc/addons/AddonManager.h [hunk 8]
```diff
@@ -69,6 +69,12 @@ namespace ADDON
     bool Init();
     void DeInit();
 
+    CAddonMgr();
```
Comment (Montellese) [inline xbmc/addons/AddonManager.h]: If we want to avoid people creating new instances of `CAddonMgr` or creating copies of it we could make `CServiceMgr` a friend class and make these methods (apart from the destructor) protected.

File: xbmc/ServiceManager.h (hunk 7 above)
Comment (FernetMenta) [inline xbmc/ServiceManager.h]: ok, no problem

File: xbmc/addons/AddonManager.h (hunk 8 above)
Comment (FernetMenta) [inline xbmc/addons/AddonManager.h]: I would keep this as simple as possible. Why trying to protect something that is very unlikely to happen? I can't think of any reason why s.o. would create another instance of CAddonMgr. If so I would be interested in that case :)

Copy and assignment constructors are only default that I would drop.

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
@tamland objections?

**Comment by tamland** [comment]
nope. go ahead

**Comment by fetzerch** [comment]
Seems like the unit tests have to be adapted as well: https://s3.amazonaws.com/archive.travis-ci.org/jobs/117262063/log.txt

**Comment by FernetMenta** [comment]
or just removed :)
not sure what they are good for.

**Comment by FernetMenta** [comment]
btw: I tried for a while but had no clue how this works.

**Comment by fetzerch** [comment]
How what works? :) Is there anyone working on the tests that we can ping for the tests?

The problem is that `CApplication::m_ServiceManager` is initialized only in `CApplication::Create()`. Since that method can't be used in the tests, they fail because `m_ServiceManager` is still a `nullptr`. I'm not sure what's the best way of fixing it though. Maybe move such initialization into a separate method in `CApplication`? Not sure if headless is still a relevant topic, but I think that would be another case requiring to decouple such initialization from creating the application.

(Moving it to `CApplication`'s constructor at least won't work, because `g_application` it a global after all and we can't guarantee that `CSpecialProtocol::m_pathMap` (used by `CAddonMgr::Init()`) is available already when `g_application` is created.)

**Comment by FernetMenta** [comment]
gtest is long outdated and has zero advantage. It should help finding issues. I am not aware of anything where it helped doing this. In the current state it just draws resources for fixing it.

**Comment by notspiff** [comment]
huh? gtest outdated? gtest is far from outdated, it's the cleanest xUnit env for c++ by far imnsho. if you mean the tests are outdated, they really aren't outdated, it's just that coverage sucks arse. fixed in https://github.com/xbmc/xbmc/pull/9419

**Comment by FernetMenta** [comment]
I don't mean the framework bur our tests. I guess that the majority of devs (like me) don't even know what exactly those tests check and how to interpret a failing test. Currently it seems that we fix the test harness after having changed the code, not the other way round.

**Comment by notspiff** [comment]
well, chicken, meet egg. that's due to badly educated devs / bad habits and not due to test driven development being broken (don't read offense into this please!). a good test suite is a better development aid than even the debugger. try it. you will (learn to) love it.

after i finally convinced my colleagues at work, bug density has gone down with several orders. granted, the things we do lend itself better to testing than a gui driven application.

**Comment by garbear** [comment]
I used the test framework when building my first attempt at a game library. it was very helpful, and i recommend developers making use of this resource.
